tools/clang/test/CodeGenHLSL/hlmatrixlower/array_load_store.ll

```diff
@@ -0,0 +1,54 @@
+; RUN: %opt %s -hlmatrixlower -S | FileCheck %s
+
```

Comment:
It would be helpful if you could add hlsl source in these .ll files as commented out blocks.

Reply:
There is no HLSL source, those are handwritten specifically to test some lowering patterns. It's difficult not to generate lots of extra noise from HLSL.

Reply:
Handwriting .ll files sounds a bit tedious to me and may be harder to maintain in case we wish to do small modifications to these files in future. Not sure how we have been generating existing .ll files for this project, but we have some internal projects for which we would almost always auto generate ll files from source. @hekota any thoughts on this? Also, by extra noise, I assume you mean additional metadata info?


  I would lean towards ll files produced from source through compilation

Reply:
It is definitively tedious, but the point of those tests are to exercise specific matrix lowering scenarios at the unit test level. When we only test with HLSL, we are doing integration tests and don't quite control the input to specific passes because compilation from HLSL will generate extra loads/stores, matrix orientation casts, etc. This is what I mean by noise.

Our testing philosophy has been to only test at the integration level so far, but it makes it difficult to isolate the kind of issue I was looking for and to detect the "two wrongs make one right" type of bugs. Having unit tests for the pass before it potentially undergoes a rewrite or important changes was @jpporto 's idea, and I think it has merit.

Reply:
I think I understood the intent, but it just looked a bit odd to me in the beginning to have that many handwritten .ll test files. However, taking a closer look at those files, I see that you are mostly testing against buggy or problematic code gen scenarios and in these cases you are ensuring that matrix lower pass handle these scenarios correctly. That makes sense..

Reply:
The intent is not actually to test buggy or problematic codegen scenarios, but rather just have focused tests for each kind of pattern that HLMatrixLower pass might have to deal with.

Reply:
Ah, I am then confused. If they are not buggy or problematic codegen scenarios, they at least appear to be special ones. In other words, below hlsl source produces initial IR which looks similar to what you have in unary_operator.ll file except that they are use regular load/store instead of rowLoad/rowStore.

**HLSL**
```
void main(out row_major int2x2 a : A)
{
 a = -a;
}
```

**IR after initial code gen:**

```
define void @main(%class.matrix.int.2.2* noalias dereferenceable(16) %a) #0 {
entry:
  %0 = call %class.matrix.int.2.2 @"dx.hl.matldst.rowLoad.%class.matrix.int.2.2 (i32, %class.matrix.int.2.2*)"(i32 2, %class.matrix.int.2.2* %a)
  %1 = call %class.matrix.int.2.2 @"dx.hl.unop.-.%class.matrix.int.2.2 (i32, %class.matrix.int.2.2)"(i32 6, %class.matrix.int.2.2 %0)
  %2 = call %class.matrix.int.2.2 @"dx.hl.matldst.rowStore.%class.matrix.int.2.2 (i32, %class.matrix.int.2.2*, %class.matrix.int.2.2)"(i32 3, %class.matrix.int.2.2* %a, %class.matrix.int.2.2 %1)
  ret void
}
```